src/screencast/wlr_screencast.c

```diff
@@ -545,28 +545,90 @@ static void wlr_remove_output(struct xdpw_wlr_output *out) {
 	free(out);
 }
 
+static void wlr_format_modifier_pair_add(struct xdpw_screencast_context *ctx,
+		uint32_t format, uint64_t modifier) {
+	struct xdpw_format_modifier_pair *fm_pair = calloc(1, sizeof(struct xdpw_format_modifier_pair));
+
+	fm_pair->fourcc = format;
+	fm_pair->modifier = modifier;
+
+	logprint(TRACE, "wlroots: format %u (%lu)", fm_pair->fourcc, fm_pair->modifier);
+
+	wl_list_insert(&ctx->format_modifier_pairs, &fm_pair->link);
+
+}
+
+static void wlr_format_modifier_pair_emtpy_list(struct xdpw_screencast_context *ctx) {
+	struct xdpw_format_modifier_pair *fm_pair, *tmp_fmp;
+	wl_list_for_each_safe(fm_pair, tmp_fmp, &ctx->format_modifier_pairs, link) {
+		wl_list_remove(&fm_pair->link);
+		free(fm_pair);
+	}
+}
+
 static void linux_dmabuf_handle_modifier(void *data,
 		struct zwp_linux_dmabuf_v1 *zwp_linux_dmabuf_v1,
 		uint32_t format, uint32_t modifier_hi, uint32_t modifier_lo) {
 	struct xdpw_screencast_context *ctx = data;
 
 	logprint(TRACE, "wlroots: linux_dmabuf_handle_modifier called");
 
-	struct xdpw_format_modifier_pair *fm_pair = calloc(1, sizeof(struct xdpw_format_modifier_pair));
-
-	fm_pair->fourcc = format;
-	fm_pair->modifier = ((((uint64_t)modifier_hi) << 32) | modifier_lo);
-
-	logprint(TRACE, "wlroots: format %u (%u)", fm_pair->fourcc, fm_pair->modifier);
-
-	wl_list_insert(&ctx->format_modifier_pairs, &fm_pair->link);
+	wlr_format_modifier_pair_add(ctx, format, ((((uint64_t)modifier_hi) << 32) | modifier_lo));
 }
 
 static const struct zwp_linux_dmabuf_v1_listener linux_dmabuf_listener = {
 	.format = noop,
 	.modifier = linux_dmabuf_handle_modifier,
 };
 
+static void linux_dmabuf_feedback_handle_main_device(void *data,
+		struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1, struct wl_array *device_arr) {
+	struct xdpw_screencast_context *ctx = data;
+
+	logprint(DEBUG, "wlroots: linux_dmabuf_feedback_handle_main_device called");
+	assert(device_arr->size == sizeof(ctx->main_device));
+	memcpy(&ctx->main_device, device_arr->data, sizeof(ctx->main_device));
+}
+
+static void linux_dmabuf_feedback_format_table(void *data,
+		struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1, int fd, uint32_t size) {
+	struct xdpw_screencast_context *ctx = data;
+
+	logprint(DEBUG, "wlroots: linux_dmabuf_feedback_format_table called");
+
+	wlr_format_modifier_pair_emtpy_list(ctx);
+
+	struct fm_entry {
+		uint32_t format;
+		uint8_t padding;
+		uint64_t modifier;
+	};
+	uint32_t n_modifiers = size / 16;
+
+	void *fm_data = mmap(NULL , size, PROT_READ, MAP_PRIVATE, fd, 0);
+
+	struct fm_entry *fm_entry = fm_data;
+	for (uint32_t i = 0; i < n_modifiers; i++) {
```

Comment:
This won't work -- the table might contain unsupported modifiers (e.g. some intended for a different device, or for a different client, or just arbitrary holes/undefined entries). We need to listen to `tranche_formats` to fill the list.

Reply:
Thanks. Implemented to go through the tranches. Just needed sth. yesterday to get dmabufs available again.